IOS/KD: Add WiiLink as WC24 service provider by default

noahpistilli opened the pull request:
As discussed, this PR has WiiLink as the default service for WiiConnect24 services by default (More are coming!)

Currently, there is a checkbox users can toggle in the config to disable or enable WiiLink. If enabled, it will reroute requests by a supportted channel to the WiiLink equivalent.

Image of the config option under `Wii`
![](https://cdn.discordapp.com/attachments/791984984700354600/1116150889489649674/Screenshot_2023-06-07_at_7.png)


TODO:
Somewhere to put the Terms of Service visibly. JMC had the idea of having it popup much like the Analytics disclaimer.

leoetlino commented:
The changes look sane - I wonder if we should use a dropdown instead of a checkbox though, just in case we decide to add more WC24 service provider in the future. Also, I think it'd be nice to have a description of what WiiLink is in the tooltip text.

leoetlino commented:
Alternatively, let's add an INI override for the server names to e.g. make local testing/development of WiiLink style services easier

noahpistilli commented:
Added an INI field as requested, as well as rename `Enable WiiLink` to `Enable WiiConnect24`. If `Enable WiiConnect24` is checked, Dolphin will load the default INI for the WiiConnect24 title. Otherwise, it will load the user's INI.

I have also added WiiLink URLs to the default INI. Nintendo Channel is also currently not in a complete state as main requests use SSL.

leoetlino commented on Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp:
```diff
@@ -217,7 +219,16 @@ NWC24::ErrorCode NetKDRequestDevice::KDDownload(const u16 entry_index,
+  // Reroute to custom server if enabled.
+  // Find domain name of current request
+  // If the URL is invalid libcurl will error out and log the error for us.
+  const std::vector<std::string> parts = SplitString(url, '/');
```
Can you check if parts contains at least 3 elements before accessing parts[2]?

leoetlino commented on Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp:
```diff
@@ -16,6 +16,7 @@
 #include "Common/Logging/Log.h"
 #include "Common/NandPaths.h"
 #include "Common/SettingsHandler.h"
+#include "Core/Config/MainSettings.h"
```
this is unused as far as I can tell

leoetlino commented on Source/Core/Core/IOS/Network/IP/Top.cpp:
```diff
@@ -1056,6 +1057,10 @@ IPCReply NetIPTopDevice::HandleGetAddressInfoRequest(const IOCtlVRequest& reques
   if (!request.in_vectors.empty() && request.in_vectors[0].size > 0)
   {
     nodeNameStr = memory.GetString(request.in_vectors[0].address, request.in_vectors[0].size);
+    if (std::optional<std::string> patch = WC24PatchEngine::GetNetworkPatch(nodeNameStr, false))
```
For the sake of making call sites more readable, consider changing the second argument to an enum class:

```cpp
// in WC24PatchEngine
enum class IsKD : bool;

std::optional<std::string> GetNetworkPatch(..., IsKD is_kd);
```

a call to GetNetworkPatch would look like this:

```cpp
WC24PatchEngine::GetNetworkPatch(nodeNameStr, IsKD{false});
```

making it immediately obvious what the boolean represents

leoetlino commented on Source/Core/DolphinQt/Settings/WiiPane.cpp:
```diff
@@ -157,6 +158,7 @@ void WiiPane::CreateMisc()
   m_main_layout->addWidget(misc_settings_group);
   m_pal60_mode_checkbox = new QCheckBox(tr("Use PAL60 Mode (EuRGB60)"));
   m_screensaver_checkbox = new QCheckBox(tr("Enable Screen Saver"));
+  m_wiilink_checkbox = new QCheckBox(tr("Enable WiiConnect24"));
```
I think the user-visible label should be "Enable WiiLink" or "Enable WiiConnect24 via WiiLink" - for consistency with what the option actually does and its name in the source code. Otherwise it's a bit misleading.

noahpistilli replied:
Chose the latter option, fits nice and is more descriptive overall.

leoetlino commented on Source/Core/Core/WC24PatchEngine.cpp:
```diff
@@ -0,0 +1,138 @@
+{
+static std::array<u64, 12> s_wc24_channels{
+    // Nintendo Channel
+    0x0001000148415445,
```
Consider adding these to Core/CommonTitles.h?

leoetlino commented:
Looks like MSVC is unhappy about operator< being used on bools; this is a side effect of the wrong overload of TryParse getting selected.

Consider adding a specialisation of TryParse for boolean enums: https://godbolt.org/z/eG1YcGee4

```cpp
// add this:
template <typename T> requires (detail::isBooleanEnum<T>())
bool TryParse(T* output) {
    bool value;
    if (!TryParse(&value))
        return false;
    
    *output = static_cast<T>(value);
    return true;
}

// and then change the integer/enum template to this:
template <typename T>
requires (std::is_integral_v<T> || (std::is_enum_v<T> && !detail::isBooleanEnum<T>()))
bool TryParse(T* output);
```

noahpistilli commented:
Just added the patches in the INI files for Nintendo Channel. Should be good to merge now.

AdmiralCurtiss commented on Data/Sys/GameSettings/HAF.ini:
```diff
@@ -0,0 +1,5 @@
+
+[WC24Patch]
+$Main
+weather.wapp.wii.com:fore.wiilink24.com:1
```
Can you please add end-of-file newlines to all your inis?

AdmiralCurtiss commented on Source/Core/Core/ConfigManager.cpp:
```diff
@@ -54,6 +54,7 @@
 #include "Core/TitleDatabase.h"
 #include "VideoCommon/HiresTextures.h"
 
+#include "Core/WC24PatchEngine.h"
```
This include should go into the block with the other Core includes. Maybe also add a newline between the Core block and the VideoCommon one while you're there.

AdmiralCurtiss commented on Source/Core/Core/IOS/Network/KD/NWC24DL.cpp:
```diff
@@ -140,4 +140,9 @@ void NWC24Dl::SetVersion(u32 version)
   m_data.header.version = Common::swap32(version);
 }
 
+u32 NWC24Dl::GetHighTitleID(u16 entry_index) const
+{
+  return Common::swap32(m_data.entries[entry_index].high_title_id);
+}
+
```
This function doesn't seem to be used anywhere?

AdmiralCurtiss commented on Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp:
```diff
@@ -217,7 +218,22 @@ NWC24::ErrorCode NetKDRequestDevice::KDDownload(const u16 entry_index,
+  if (std::optional<std::string> patch =
+          WC24PatchEngine::GetNetworkPatch(parts[2], WC24PatchEngine::IsKD{true}))
+  {
+    url = ReplaceAll(url, parts[2], patch.value());
```
This seems too broad, this would also replace instances outside the domain part.

AdmiralCurtiss commented on Source/Core/Core/WC24PatchEngine.cpp:
```diff
@@ -0,0 +1,136 @@
+{
+  for (const NetworkPatch& patch : s_patches)
+  {
+    if (source.find(patch.source) != std::string::npos && patch.is_kd != IsKD{true})
+      // We found the correct patch, return it!
+      return ReplaceAll(std::string{source}, patch.source, patch.replacement);
```
This also seems too aggressive. You'd be replacing any instances outside the Host header, too.

AdmiralCurtiss replied:
This on the other hand is still clearly wrong. You're targetting the `Host:` part of the HTTP header here, yeah? https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages In that case, you should also only search in that area of the request. Basically, something like (untested if this actually works, you might want to write a unit test for the logic because it seems prone to off-by-one errors...)

```
  size_t pos = 0;
  while (pos < source.size())
  {
    if (source[pos] == '\n')  // end of header
      break;

    if (source.substr(pos).starts_with("Host:"))
    {
      size_t end_of_line = source.find('\n', pos);
      std::string_view domain =
          source.substr(pos + 5, end_of_line == std::string_view::npos ? std::string_view::npos :
                                                                         (end_of_line - pos - 5));
      for (const NetworkPatch& patch : s_patches)
      {
        if (domain.trim() == patch.source)
        {
          return source.substr(0, pos) + "Host: " + patch.replacement +
                 (end_of_line == std::string_view::npos ? "" : source.substr(end_of_line));
        }
      }
    }
    else
    {
      size_t end_of_line = source.find('\n', pos);
      if (end_of_line == std::string_view::npos)
        break;
      pos = end_of_line + 1;
    }
  }
  return std::nullopt;
```

noahpistilli replied:
This example was a big help. With some testing and looking at [RFC2616](https://www.rfc-editor.org/rfc/rfc2616#section-2.2), all HTTP messages have CRLF line endings, so I opted for checking for that instead of just `\n`. Other than that and the return line, it works perfectly with my testing in the Nintendo Channel.

AdmiralCurtiss commented:
Aside from those this seems fine. I'll grant that it's not particularly likely that those domains would appear anywhere outside the specific places you're expecting them, but it still doesn't seem a great idea to run a blanket search-and-replace when you could be more targetted to what you're actually expecting.

AdmiralCurtiss commented on Source/Core/Core/IOS/Network/KD/NetKDRequest.cpp:
```diff
@@ -217,7 +218,23 @@ NWC24::ErrorCode NetKDRequestDevice::KDDownload(const u16 entry_index,
+          WC24PatchEngine::GetNetworkPatch(parts[2], WC24PatchEngine::IsKD{true}))
+  {
+    const size_t index = url.find(parts[2]);
+    url.replace(index, parts[2].size(), patch.value());
```
This is *better* but I feel you could still trick this if you really tried to...

It's probably *fine* given typical input but if you feel motivated you might want to actually extract the host part of the URI based on https://www.rfc-editor.org/rfc/rfc3986 and just target that part.

noahpistilli replied:
If you *really* want me to I can fix it, otherwise I might just defer it to another time.

AdmiralCurtiss replied:
Eh... like I said, if you feel motivated. You'd have to actively try to abuse this, and to questionable effect.

AdmiralCurtiss commented on Source/Core/Core/Config/MainSettings.cpp:
```diff
@@ -241,6 +241,7 @@ const Info<bool> MAIN_ALLOW_SD_WRITES{{System::Main, "Core", "WiiSDCardAllowWrit
 const Info<bool> MAIN_ENABLE_SAVESTATES{{System::Main, "Core", "EnableSaveStates"}, false};
 const Info<bool> MAIN_REAL_WII_REMOTE_REPEAT_REPORTS{
     {System::Main, "Core", "RealWiiRemoteRepeatReports"}, true};
+const Info<bool> MAIN_WII_WIILINK_ENABLE{{System::Main, "Core", "EnableWiiLink"}, true};
```
This should probably default to off.

AdmiralCurtiss commented on Source/Core/Core/WC24PatchEngine.cpp:
```diff
@@ -0,0 +1,159 @@
+  size_t pos{};
+  while (pos < source.size())
+  {
+    if (source.substr(pos).starts_with("Host:"))
+    {
+      size_t end_of_line = source.find("\r\n", pos);
+      std::string_view domain =
+          source.substr(pos + 6, end_of_line == std::string_view::npos ? std::string_view::npos :
```
The `size_t end_of_line = source.find("\r\n", pos);` can be pulled out of the branch.

There's a mismatch here between the `starts_with("Host:")` and `.substr(pos + 6`, you're dropping a byte here. I'm not sure whether HTTP forces a space between the `:` and the start of the data? Might be okay to just change to `starts_with("Host: ")`?

Adding a few `const`s to variables here wouldn't hurt either.

AdmiralCurtiss commented on Source/Core/Core/WC24PatchEngine.cpp:
```diff
@@ -0,0 +1,159 @@
+  if (!IsWC24Channel())
+    return;
+
+  Common::IniFile ini;
+  // If WiiLink is enabled then we load the Default Ini as that has the WiiLink URLs.
+  if (Config::Get(Config::MAIN_WII_WIILINK_ENABLE))
+    ini = sconfig.LoadDefaultGameIni();
+  else
+    ini = sconfig.LoadLocalGameIni();
```
This logic here also feels a bit jank now that I think about it... I would expect user patches to apply in addition to the system ones, regardless of if the WiiLink option is enabled or not. Though I'd be okay with that being a follow-up by either you or someone else, I'm not expecting every user to suddenly start writing URL replacements into their INIs.

noahpistilli replied:
The logic I was trying to get at here is that the WiiLink URL's will be contained within the INI's we pack. If Enable WiiLink is off, it would allow for users to put in their own replacement URL if they please. I just didn't want any conflicts between the INI's URL's

AdmiralCurtiss commented on Source/Core/Core/WC24PatchEngine.cpp:
```diff
@@ -0,0 +1,158 @@
+                                                                     source.substr(end_of_line));
+        }
+      }
+    }
```
Once the Host header is found but none of the patches match you can break out of the loop and return std::nullopt.

AdmiralCurtiss replied:
Oh and I just realized... the patches have an `enabled` field that isn't checked here, I assume it should be?

AdmiralCurtiss commented on Source/Core/Core/WC24PatchEngine.cpp:
```diff
@@ -0,0 +1,158 @@
+    }
+  }
+
+  ReadEnabledAndDisabled(ini, "WC24Patch", &s_patches);
```
Does this actually do anything? I see no `WC24Patch_Enabled` section in the ini, and no way to enable or disable the individual patches either.

noahpistilli replied:
Thank you so much for this catch, I must have deleted the `WC24Patch_Enabled` while doing something. It is re-added now.

AdmiralCurtiss commented on Source/Core/Core/WC24PatchEngine.cpp:
```diff
@@ -0,0 +1,158 @@
+    {
+      const std::string_view domain =
+          source.substr(pos + 6, end_of_line == std::string_view::npos ? std::string_view::npos :
+                                                                         (end_of_line - pos - 5));
```
This needs to be `end_of_line - pos - 6` so that you don't grab the `\r`...

AdmiralCurtiss reviewed (approved):
Untested, but this looks reasonable enough to me now.

noahpistilli commented on Source/Core/Core/IOS/Network/Socket.cpp:
```diff
@@ -581,6 +582,12 @@ void WiiSocket::Update(bool read, bool write, bool except)
+          if (patch)
+          {
+            data = patch->c_str();
+            BufferInSize = static_cast<u32>(patch->size()) - 2;
```
Upon reviewing payload sizes and how they look, all Nintendo Channel payloads were `data_size - 2`. This is because the literal `FR` was at the end of the string, not sure what it is for but it should not be sent through the socket.

AdmiralCurtiss replied:
Oh, duh, yeah that's not a nullterminated string so we're treating it incorrectly. That explains your corruption. Please change it to

```
          const std::optional<std::string> patch = WC24PatchEngine::GetNetworkPatchByPayload(std::string_view(data, data + BufferInSize));
          if (patch)
          {
            data = patch->c_str();
            BufferInSize = static_cast<u32>(patch->size()) - 2;
          }
```

and see if it's correct then.

noahpistilli replied:
Works like a charm.
